sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/PushDownLeftSemiAntiJoin.scala

```diff
@@ -105,16 +114,25 @@ object PushDownLeftSemiAntiJoin extends Rule[LogicalPlan] with PredicateHelper {
         }
 
         // Check if the remaining predicates do not contain columns from the right
-        // hand side of the join. Since the remaining predicates will be kept
-        // as a filter over window, this check is necessary after the left semi
-        // or left anti join is moved below window. The reason is, for this kind
-        // of join, we only output from the left leg of the join.
+        // hand side of the join. In case of LeftSemi join, since remaining predicates
+        // will be kept as a filter over aggregate, this check is necessary after the left semi join
+        // is moved below aggregate. The reason is, for this kind of join, we only output from the
+        // left leg of the join.
         val rightOpColumns = AttributeSet(stayUp.toSet).intersect(rightOp.outputSet)
 
         if (pushDown.nonEmpty && rightOpColumns.isEmpty) {
           val predicate = pushDown.reduce(And)
           val newPlan = w.copy(child = Join(w.child, rightOp, joinType, Option(predicate), hint))
-          if (stayUp.isEmpty) newPlan else Filter(stayUp.reduce(And), newPlan)
+          if (stayUp.isEmpty) {
+            newPlan
+          } else {
+            // In case of left anti join, the join is pushed down when the entire join condition
+            // is eligible to be pushed down to preserve the semantics of left anti join.
+            joinType match {
+              case LeftSemi => Filter(stayUp.reduce(And), newPlan)
+              case _ => join
```

Comment:
how about
```
joinType match {
  // In case of left-semi join, ...
  case LeftSemi => Filter(stayUp.reduce(And), newPlan)
  // In case of left-anti join, ...
  case _ => join
```

Reply:
@cloud-fan I think it would look a bit crowded :-). The actual code may look buried within the comment. Let me give it a try and we can a take a call after ..

Reply:
@cloud-fan Just made the change... Let me know if looks okay to you.